tests/data/semparse/type_declarations/type_declaration_test.py

```diff
@@ -8,3 +11,57 @@ def test_basic_types_conflict_on_names(self):
         type_a = types.NamedBasicType("A")
         type_b = types.NamedBasicType("B")
         assert type_a.resolve(type_b) is None
+
+    def test_valid_actions(self):
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <e,<r,<d,r>>>
+        type_signatures = {'F': ComplexType(type_e, ComplexType(type_r, ComplexType(type_d, type_r)))}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<e,<r,<d,r>>>"] == {"<e,<r,<d,r>>> -> sample_function"}
+        assert valid_actions["<r,<d,r>>"] == {"<r,<d,r>> -> [<e,<r,<d,r>>>, e]"}
+        assert valid_actions["<d,r>"] == {"<d,r> -> [<r,<d,r>>, r]"}
+        assert valid_actions["r"] == {"r -> [<d,r>, d]"}
+
+    def test_valid_actions_with_placeholder_type(self):
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <#1,#1>
+        type_signatures = {'F': types.IdentityType(types.ANY_TYPE, types.ANY_TYPE)}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<#1,#1>"] == {"<#1,#1> -> sample_function"}
+        assert valid_actions["e"] == {"e -> [<#1,#1>, e]"}
+        assert valid_actions["r"] == {"r -> [<#1,#1>, r]"}
+        assert valid_actions["d"] == {"d -> [<#1,#1>, d]"}
+
+    def test_valid_actions_with_any_type(self):
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <#1,r>
+        type_signatures = {'F': ComplexType(types.ANY_TYPE, type_r)}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<d,r>"] == {"<d,r> -> sample_function"}
```

Comment:
Related to the above, why are these instantiated to basic types while the placeholders aren't?  Why not just leave it as `<ANY,r> -> sample_function`?

Reply:
I do not foresee defining functions with `ANY_TYPE` like I do in this test. We should just use `PlaceholderType` instead. The purpose of this test is to check whether `ANY_TYPE` get substituted by every possible type, and the idea is to simulate an intermediate step while building up a tree to get the actions for a `PlaceholderType`.

In general, I think of `ANY_TYPE`s as those that get resolved during parsing, and placeholders as those that do not. This means that we'll have placeholder types only within terminal productions.

Reply:
Added a comment within this test.